jetty-http3/http3-qpack/src/main/java/org/eclipse/jetty/http3/qpack/QpackDecoder.java

```diff
@@ -0,0 +1,276 @@
+//
+// ========================================================================
+// Copyright (c) 1995-2021 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.http3.qpack;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.eclipse.jetty.http.HttpField;
+import org.eclipse.jetty.http.MetaData;
+import org.eclipse.jetty.http3.qpack.internal.QpackContext;
+import org.eclipse.jetty.http3.qpack.internal.instruction.InsertCountIncrementInstruction;
+import org.eclipse.jetty.http3.qpack.internal.instruction.SectionAcknowledgmentInstruction;
+import org.eclipse.jetty.http3.qpack.internal.parser.DecoderInstructionParser;
+import org.eclipse.jetty.http3.qpack.internal.parser.EncodedFieldSection;
+import org.eclipse.jetty.http3.qpack.internal.table.DynamicTable;
+import org.eclipse.jetty.http3.qpack.internal.table.Entry;
+import org.eclipse.jetty.http3.qpack.internal.table.StaticTable;
+import org.eclipse.jetty.http3.qpack.internal.util.NBitIntegerParser;
+import org.eclipse.jetty.util.BufferUtil;
+import org.eclipse.jetty.util.component.Dumpable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Qpack Decoder
+ * <p>This is not thread safe and may only be called by 1 thread at a time.</p>
+ */
+public class QpackDecoder implements Dumpable
+{
+    public static final Logger LOG = LoggerFactory.getLogger(QpackDecoder.class);
+
+    private final List<Instruction> _instructions = new ArrayList<>();
+    private final Instruction.Handler _handler;
+    private final QpackContext _context;
+    private final DecoderInstructionParser _parser;
+    private final List<EncodedFieldSection> _encodedFieldSections = new ArrayList<>();
+    private final NBitIntegerParser _integerDecoder = new NBitIntegerParser();
+    private final int _maxHeaderSize;
+    private MetaData _metaData;
```

Comment:
@lachlan-roberts 
A `Decoder` can decode `MetaData` for multiple streams, so why is there a single `_metaData` field?  I'm guessing this is the current meta data field being built?  But are we sure that we can't be called like:
```java
    decoder.decode(streamA, bufferWithFirstHalfOfMetaDataA, myHandler);
    decoder.decode(streamB, bufferWithFirstHalfOfMetaDataB, myHandler);
    decoder.decode(streamA, bufferWithSecondHalfOfMetaDataA, myHandler);
    decoder.decode(streamB, bufferWithSecondHalfOfMetaDataB, myHandler);
```
If this can't happen, then at the very least this needs to be documented, but I'd also rather we remembered the streamId in progress and threw an ISE if we are called with another stream when we have not finished the first.